src/yara.ml

```diff
@@ -82,16 +82,69 @@ let finalize () =
 let finalize_exn () = Yr.unwrap @@ finalize ()
 
 module Rule = struct
-  type t = Yara_c.Yr_rule.t ptr
-
-  let get_identifier rule =
-    let rule = !@rule in
-    getf rule Yara_c.Yr_rule.identifier
+  type metadata =
+    | Int of int64
+    | Bool of bool
+    | String of string
+
+  type t = {
+    identifier : string;
+    namespace : string;
+    metadata : (string * metadata) list;
+  }
+
+  let collect_metadata metas =
+    let q = Queue.create () in
+    let ident meta = getf meta Yara_c.Yr_meta.identifier in
+    let rec loop metas =
+      if not (is_null metas) then (
+        let meta = !@metas in
+        let flags = getf meta Yara_c.Yr_meta.flags in
+        let last_in_rule = flags = Yara_c.Constants.meta_flags_last_in_rule in
+        let ident = ident meta in
+        match getf meta Yara_c.Yr_meta.type_ with
+        | Yara_c.Meta_type.Integer ->
+          let v = getf meta Yara_c.Yr_meta.integer in
+          Queue.add (ident, Int v) q;
+          if not last_in_rule then loop (metas +@ 1)
+        | String ->
+          let v = getf meta Yara_c.Yr_meta.string in
+          Queue.add (ident, String v) q;
+          if not last_in_rule then loop (metas +@ 1)
+        | Boolean ->
+          let v = getf meta Yara_c.Yr_meta.integer in
+          Queue.add
+            ( ident,
+              Bool
+                ( if v > 0L then
+                  true
+                else
+                  false
+                )
+            )
+            q;
+          if not last_in_rule then loop (metas +@ 1)
+        | Unknown _ -> loop (metas +@ 1)
+      )
+    in
+    loop metas;
+    List.of_seq (Queue.to_seq q)
 
-  let get_namespace rule =
-    let rule = !@rule in
+  let from_ptr ptr =
+    let rule = !@ptr in
     let ns = !@(getf rule Yara_c.Yr_rule.namespace) in
```

Comment:
The pointer dereference should be safe as this is only called from the scan callback where we should have a `YR_RULE`. That said, we can also be defensive, check for null pointers and raise a more specific exception here.

Same goes for `ns`. I believe if a user doesn't specify a namespace yara will use a `default` namespace and as a result we shouldn't be in a situation where this pointer is null?

Reply:
Added a check for nullptrs